src/hotspot/share/utilities/elfFile.cpp

```diff
@@ -348,4 +302,1408 @@ ElfStringTable* ElfFile::get_string_table(int index) {
+  if (debug_filename == nullptr) {
+    return nullptr;
+  }
+
+  mfd.set_position(shdr.sh_offset);
+  if (!mfd.read(debug_filename, shdr.sh_size)) {
+    return nullptr;
+  }
+  return debug_filename;
+}
+
+// Try to load the dwarf file from the same directory as the library file.
+bool ElfFile::load_dwarf_file_from_same_directory(DwarfFilePath& dwarf_file_path) {
+  dwarf_file_path.set_filename_after_last_slash();
+  return open_valid_debuginfo_file(dwarf_file_path);
+}
+
+// Try to load the dwarf file from a user specified path in environmental variable _JVM_DWARF_PATH.
+bool ElfFile::load_dwarf_file_from_env_var_path(const DwarfFilePath& dwarf_file_path) {
+  const char* dwarf_path = ::getenv("_JVM_DWARF_PATH");
+  if (dwarf_path != nullptr) {
+    log_develop_debug(dwarf)("_JVM_DWARF_PATH: %s", dwarf_path);
+    const char* filename = dwarf_file_path.filename();
+    return (load_dwarf_file_from_env_path_folder(dwarf_path, "/lib/server/", filename)
+            || load_dwarf_file_from_env_path_folder(dwarf_path, "/lib/", filename)
+            || load_dwarf_file_from_env_path_folder(dwarf_path, "/bin/", filename)
+            || load_dwarf_file_from_env_path_folder(dwarf_path, "/", filename));
+  }
+  return false;
+}
+
+bool ElfFile::load_dwarf_file_from_env_path_folder(const char* env_path, const char* folder, const char* filename) {
+  const size_t dwarf_filepath_buf_len = strlen(env_path) + strlen(folder) + strlen(filename) + 2;
+  char* dwarf_filepath_buf = NEW_RESOURCE_ARRAY(char, dwarf_filepath_buf_len);
+  if (dwarf_filepath_buf == nullptr) {
+    log_develop_info(dwarf)("Failed to allocate path buffer");
+    return false;
+  }
+
+  DwarfFilePath dwarf_file_path(filename, dwarf_filepath_buf, dwarf_filepath_buf_len);
+  dwarf_file_path.set(env_path);
+  dwarf_file_path.append(folder);
+  dwarf_file_path.append(filename);
+  return open_valid_debuginfo_file(dwarf_file_path);
+}
+
+// Try to load the dwarf file from a subdirectory named .debug within the directory of the library file.
+bool ElfFile::load_dwarf_file_from_debug_sub_directory(DwarfFilePath& dwarf_file_path) {
+  dwarf_file_path.set_after_last_slash(".debug/");
+  dwarf_file_path.append(dwarf_file_path.filename());
+  return open_valid_debuginfo_file(dwarf_file_path);
+}
+
+// Try to load the dwarf file from /usr/lib/debug + the full pathname.
+bool ElfFile::load_dwarf_file_from_usr_lib_debug(DwarfFilePath& dwarf_file_path) {
+  dwarf_file_path.set(usr_lib_debug_directory());
+  dwarf_file_path.append(_filepath);
+  dwarf_file_path.set_filename_after_last_slash();
+  return open_valid_debuginfo_file(dwarf_file_path);
+}
+
+bool ElfFile::read_section_header(const char* name, Elf_Shdr& hdr) const {
+  if (_shdr_string_table == nullptr) {
+    // Section header string table should be loaded
+    return false;
+  }
+  size_t len = strlen(name) + 1;
+  ResourceMark rm;
+  char* buf = NEW_RESOURCE_ARRAY(char, len);
+  if (buf == nullptr) {
+    return false;
+  }
```

Comment:
Thomas's comment caught my attention in the email. NEW_RESOURCE_ARRAY aborts the VM on OOM. Use NEW_RESOURCE_ARRAY_RETURN_NULL if you want to continue.

Reply:
As I wrote in another comment, I'd rather we avoid RA altogether since it relies on Thread::current(), and we want to see callstacks even with Thread::current==NULL.